### PR TITLE
Outputs from Test tasks are exposed as variants to dependency management

### PR DESCRIPTION
**Pull request by tresat**
Fixes 18791

First draft of work for the Associated Spec for review.

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java [hunk 1]
```diff
@@ -426,6 +430,33 @@ private void configureConfigurations(Project project, JavaPluginExtension extens
...
+        variant.attributes(attributes -> {
+            attributes.attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.class, Usage.VERIFICATION));
+            attributes.attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category.class, Category.SOURCES));
+            attributes.attribute(Sources.SOURCES_ATTRIBUTE, objects.named(Sources.class, Sources.ALL_SOURCE_DIRS));
+        });
+
+        // TODO: 18791 - This is a hack until we have a proper way of injecting ObjectFactory
+        ((DefaultConfigurationPublications) variant.getOutgoing()).setObjectFactory(objects);
+
+        variant.getOutgoing().artifacts(main.getAllSource().getSourceDirectories().getElements().flatMap(e -> project.provider(() -> e)), artifact -> {
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java:455]: What's the extra levels of indirection here for?  getElements returns a provider, flatMap returns a provider and project.provider() creates another new provider.  Can we not just pass `getElements()` here?

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle cancel

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle clean

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/LazyAttributesIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            sampleProperty.set("original value")
+
+            configurations {
+                sample {
+                    visible = false
+                    canBeResolved = false
+                    canBeConsumed = true
+
+                    attributes {
+                        attribute(Usage.USAGE_ATTRIBUTE, sampleProperty.flatMap(value -> project.provider(() -> objects.named(Usage, value))))
```
Comment (big-guy) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/LazyAttributesIntegrationTest.groovy:40]: I think we can do something simpler here...

```
sampleProperty.map(value -> objects.named(Usage, value))
```

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/LazyAttributesIntegrationTest.groovy [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+            configurations {
+                sample {
+                    visible = false
+                    canBeResolved = false
+                    canBeConsumed = true
+                    extendsFrom(configurations.implementation)
+
+                    attributes {
+                        attribute(Usage.USAGE_ATTRIBUTE, testProvider.flatMap(value -> project.provider(() -> objects.named(Category, value))))
```
Comment (big-guy) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/LazyAttributesIntegrationTest.groovy:105]: Same as above

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/LazyAttributesIntegrationTest.groovy [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            """.stripIndent()
+
+        expect:
+        succeeds "outgoingVariants"
+        outputContains("""
+            --------------------------------------------------
+            Variant sample
+            --------------------------------------------------
+            Capabilities
+                - :${getTestDirectory().getName()}:unspecified (default capability)
```
Comment (big-guy) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/LazyAttributesIntegrationTest.groovy:55]: Instead of doing this, I would set a root project name and use that here

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java [hunk 5]
```diff
@@ -72,42 +88,85 @@ public String toString() {
...
+            actualValueType = provider.getType();
+        } else if (!Provider.class.isAssignableFrom(value.getClass())) {
+            actualValueType = value.getClass();
+        } else {
+            actualValueType = null;
+        }
+
+        if (null != actualValueType) {
+            if (!attribute.getType().isAssignableFrom(actualValueType)) {
+                throw new IllegalArgumentException("Unexpected type for attribute: '" + attribute.getName() + "'. Attribute type: " + attribute.getType().getName() + " did not match actual type: " + actualValueType.getName());
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java:118]: I think this doesn't quite read the right way. We say "Unexpected type" and refer to the "actual type" but the message makes it sound like the type we found for the value is the correct type.

I would leave this in terms of expected type, e.g.:

> Unexpected type for attribute 'org.gradle.usage'. Attribute type 'Integer' did not match expected type 'String'.

Where Integer is the type of the value passed in and String is the type the attribute expects.

File: subprojects/docs/src/samples/build-organization/structuring-software-projects/kotlin/build-logic/commons/src/main/kotlin/com.example.jacoco.gradle.kts [hunk 6]
```diff
@@ -15,9 +15,9 @@ configurations.create("transitiveSourcesElements") {
     isCanBeConsumed = true
     extendsFrom(configurations.implementation.get())
     attributes {
-        attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.JAVA_RUNTIME))
-        attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category.DOCUMENTATION))
-        attribute(DocsType.DOCS_TYPE_ATTRIBUTE, objects.named("source-folders"))
+        attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage::class.java, Usage.JAVA_RUNTIME))
+        attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category::class.java, Category.DOCUMENTATION))
+        attribute(DocsType.DOCS_TYPE_ATTRIBUTE, objects.named(DocsType::class.java, "source-folders"))
```
Comment (big-guy) [inline subprojects/docs/src/samples/build-organization/structuring-software-projects/kotlin/build-logic/commons/src/main/kotlin/com.example.jacoco.gradle.kts:20]: Can we update these samples to rely on the new variants?

File: subprojects/docs/src/samples/java/jvm-multi-project-with-code-coverage/kotlin/buildSrc/src/main/kotlin/myproject.jacoco-aggregation.gradle.kts [hunk 7]
```diff
@@ -14,9 +14,9 @@ val sourcesPath: Configuration by configurations.creating {
     isCanBeConsumed = false
     extendsFrom(configurations.implementation.get())
     attributes {
-        attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.JAVA_RUNTIME))
-        attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category.DOCUMENTATION))
-        attribute(DocsType.DOCS_TYPE_ATTRIBUTE, objects.named("source-folders"))
+        attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage::class.java, Usage.JAVA_RUNTIME))
```
Comment (big-guy) [inline subprojects/docs/src/samples/java/jvm-multi-project-with-code-coverage/kotlin/buildSrc/src/main/kotlin/myproject.jacoco-aggregation.gradle.kts:17]: This is the Kotlin version of the sample

File: subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoPluginIntegrationTest.groovy [hunk 8]
```diff
@@ -144,5 +144,159 @@ class JacocoPluginIntegrationTest extends AbstractIntegrationSpec {
...
+    def "jacoco plugin adds outgoing variants for default test suite"() {
+        expect:
+        succeeds "outgoingVariants"
+
+        outputContains("""
+            --------------------------------------------------
+            Variant coverageDataElementsForTest
+            --------------------------------------------------
+            Capabilities
+                - :${getTestDirectory().getName()}:unspecified (default capability)
```
Comment (big-guy) [inline subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoPluginIntegrationTest.groovy:158]: Please set a rootProject.name for this instead

File: subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/JvmTestSuitePluginIntegrationTest.groovy [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+            Attributes
+                - org.gradle.category      = documentation
+                - org.gradle.docstype      = test-results-bin
+                - org.gradle.targetname    = test
+                - org.gradle.testsuitename = test
+                - org.gradle.testsuitetype = unit-tests
+                - org.gradle.usage         = verification
+
+            Artifacts
+                - build${File.separator}test-results${File.separator}test${File.separator}binary${File.separator}results.bin (artifactType = binary)
```
Comment (big-guy) [inline subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/JvmTestSuitePluginIntegrationTest.groovy:59]: Ack, these things are making me go cross-eyed.

WDYT about something like this to improve readability?
https://github.com/gradle/gradle/commit/b1be20f871f819bdd94c5a2c7b94223105292df9

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmTestSuite.java [hunk 10]
```diff
@@ -66,6 +68,23 @@
...
+     */
+    Property<String> getTestType();
+
+    /**
+     * Set the test type for this test suite.
+     *
+     * @param testType the test type for this test suite
+     * @since 7.4
+     */
+    void setTestType(String testType);
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/JvmTestSuite.java:86]: We shouldn't add setters for Provider API based properties

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java [hunk 11]
```diff
@@ -100,6 +101,7 @@ private TestingFramework(Frameworks framework, String version) {
     private final JvmComponentDependencies dependencies;
     private boolean attachedDependencies;
     private final Action<Void> attachDependencyAction;
+    private final Property<String> testType = getObjectFactory().property(String.class).convention(TestType.UNIT_TESTS);
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java:104]: This convention needs to be set in the plugin. I think this also lets you get rid of the field and getter too.

**Review by big-guy — CHANGES_REQUESTED** [review]
I've reviewed the easy parts and I still have another ~10 files I haven't looked at fully, but here's a first round of some things I noticed.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java [hunk 12]
```diff
@@ -162,6 +172,34 @@ public void artifact(Object notation, Action<? super ConfigurablePublishArtifact
         configureAction.execute(publishArtifact);
     }
 
+    @Override
```
Comment (DPUkyle) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java:175]: This method is nearly identical to `artifacts(Provider, Action)`. I wonder if it's possible for `artifacts(Provider)` to invoke the other, passing an empty, no-op or null Action?

```
public void artifacts(Provider<? extends Iterable<? extends Object>> provider) {
  artifacts(provider, () -> { /* no-op*/ });
}
```

File: subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/DefaultIvyArtifactName.java [hunk 13]
```diff
@@ -33,12 +33,18 @@
...
-        if (name == null) {
-            name = publishArtifact.getFile().getName();
+        // TODO: 18791 There must be a better way to handle the failure of JacocoPluginIntegrationTest#"reports miss configuration of destination file"
+        // after adding the destination file as an artifact in JacocoPlugin - that test uses a provider { null } for the destination file.
+        try {
+            String name = publishArtifact.getName();
+            if (name == null) {
+                name = publishArtifact.getFile().getName();
+            }
+            String classifier = GUtil.elvis(publishArtifact.getClassifier(), null);
```
Comment (DPUkyle) [inline subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/DefaultIvyArtifactName.java:43]: This is so odd - I wonder why the original author didn't use `GUtil#elvis` to assign `name`?

So it's not your fault but I suggest removing the usage of `GUtil`, since we also don't add references to `g.l.Closure` when we can help it.

File: subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/DefaultIvyArtifactName.java [hunk 14]
```diff
@@ -33,12 +33,18 @@
...
+            return new DefaultIvyArtifactName(name, publishArtifact.getType(), publishArtifact.getExtension(), classifier);
+        } catch (Exception e) {
+            return null;
         }
```
Comment (DPUkyle) [inline subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/DefaultIvyArtifactName.java:47]: Yikes, catch-and-eat is scary. Are we sure returning null is ok here?

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublicationsTest.groovy [hunk 15]
```diff
@@ -201,4 +207,37 @@ class DefaultConfigurationPublicationsTest extends Specification {
...
     }
+
+    def "can declare outgoing artifacts using lazy provider for configuration"() {
+        given:
+        artifactNotationParser.parseNotation(_ as DefaultFileSystemLocation) >> { args ->
+            def mockArtifact = Mock(ConfigurablePublishArtifact)
+            mockArtifact.getName() >> ((DefaultFileSystemLocation) args[0]).getAsFile().getName()
+            return mockArtifact
+        }
+        publications.setObjectFactory(TestUtil.objectFactory())
```
Comment (DPUkyle) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublicationsTest.groovy:218]: Maybe a note here too that we indent to remove this mutator in favor of injection.

File: subprojects/docs/src/snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/kotlin/producer/build.gradle.kts [hunk 16]
```diff
@@ -7,15 +7,13 @@ val instrumentedJars by configurations.creating {
...
+        attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category::class.java, Category.LIBRARY))
+        attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage::class.java, Usage.JAVA_RUNTIME))
+        attribute(Bundling.BUNDLING_ATTRIBUTE, objects.named(Bundling::class.java, Bundling.EXTERNAL))
         attribute(TargetJvmVersion.TARGET_JVM_VERSION_ATTRIBUTE, JavaVersion.current().majorVersion.toInt())
-        attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, namedAttribute("instrumented-jar"))
+        attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, objects.named(LibraryElements::class.java, "instrumented-jar"))
     }
 }
-
-inline fun <reified T: Named> Project.namedAttribute(value: String) = objects.named(T::class.java, value)
```
Comment (DPUkyle) [inline subprojects/docs/src/snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/kotlin/producer/build.gradle.kts:18]: Maybe we should ship our own extension method on Project?

File: subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoPluginIntegrationTest.groovy [hunk 17]
```diff
@@ -144,5 +144,159 @@ class JacocoPluginIntegrationTest extends AbstractIntegrationSpec {
...
+                }
+            }
+
+            dependencies {
+                coverageData project
+            }
+
+            def testResolve = tasks.register('testResolve') {
+                doLast {
+                    assert coverageDataConfig.getResolvedConfiguration().getFiles()*.getName() == [test.jacoco.destinationFile.name]
```
Comment (DPUkyle) [inline subprojects/jacoco/src/integTest/groovy/org/gradle/testing/jacoco/plugins/JacocoPluginIntegrationTest.groovy:236]: Nit: be groovier using property syntax: `assert coverageDataConfig.resolvedConfiguration.files*.name == [test.jacoco.destinationFile.name]`

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java [hunk 18]
```diff
@@ -156,6 +158,16 @@ public DefaultJvmTestSuite(String name, ConfigurationContainer configurations, D
...
 
+    @Override
+    public Property<String> getTestType() {
+        return testType;
+    }
+
+    @Override
+    public void setTestType(String testType) {
+        this.testType.set(testType);
+    }
```
Comment (DPUkyle) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java:169]: No need for a setter here either

**Review by DPUkyle — COMMENTED** [review]
Overall looks great. The one surprise was `TEST_TYPE_ATTRIBUTE`; can we discuss?

I imagine the idea is to group many targets into one bucket, but I also worry about the mapping of non-unit suites and targets to a `org.gradle.testsuitetype` attribute.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/LazyAttributesIntegrationTest.groovy (hunk 2 above)
Comment (tresat) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/LazyAttributesIntegrationTest.groovy:40]: I think I needed this at one point in the test-fix cycle, but the simpler version works now.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java [hunk 19]
```diff
@@ -81,9 +86,14 @@ public DefaultConfigurationPublications(DisplayName displayName,
         this.fileCollectionFactory = fileCollectionFactory;
         this.attributesFactory = attributesFactory;
         this.domainObjectCollectionFactory = domainObjectCollectionFactory;
+        this.objectFactory = objectFactory;
         this.attributes = attributesFactory.mutable(parentAttributes);
     }
 
+    public void setObjectFactory(ObjectFactory objectFactory) {
```
Comment (big-guy) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java:93]: Once https://github.com/gradle/gradle/pull/18921 is in, please rebase/merge this PR and remove this

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java [hunk 20]
```diff
@@ -90,6 +102,44 @@ public void apply(Project project) {
...
+        variant.setCanBeConsumed(true);
+
+        final ObjectFactory objects = project.getObjects();
+        variant.attributes(attributes -> {
+            attributes.attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage.class, Usage.VERIFICATION));
+            attributes.attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category.class, Category.DOCUMENTATION));
+            attributes.attribute(DocsType.DOCS_TYPE_ATTRIBUTE, objects.named(DocsType.class, DocsType.JACOCO_COVERAGE));
+            attributes.attribute(Verification.TEST_SUITE_NAME_ATTRIBUTE, objects.named(Verification.class, suite.getName()));
+            attributes.attribute(Verification.TARGET_NAME_ATTRIBUTE, objects.named(Verification.class, suite.getName()));
+            attributes.attribute(TestType.TEST_TYPE_ATTRIBUTE, suite.getTestType().flatMap(tt -> project.provider(() -> objects.named(TestType.class, tt))));
```
Comment (big-guy) [inline subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java:134]: I think this can be simplified

File: subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java [hunk 21]
```diff
@@ -90,6 +102,44 @@ public void apply(Project project) {
...
+        });
+
+        variant.getOutgoing().artifact(target.getTestTask().flatMap(task -> project.provider(() -> task.getExtensions().getByType(JacocoTaskExtension.class).getDestinationFile())), artifact -> {
```
Comment (big-guy) [inline subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPlugin.java:137]: Hrm, I think this can be simplified, but I'm not sure how.

File: subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/JvmTestSuitePluginIntegrationTest.groovy [hunk 22]
```diff
@@ -0,0 +1,267 @@
...
+                attributes {
+                    attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage, Usage.VERIFICATION))
+                    attribute(Category.CATEGORY_ATTRIBUTE, objects.named(Category, Category.DOCUMENTATION))
+                    attribute(DocsType.DOCS_TYPE_ATTRIBUTE, objects.named(DocsType, DocsType.TEST_RESULTS))
+                }
+            }
+
+            def testResolve = tasks.register('testResolve') {
+                doLast {
+                    assert testDataConfig.getResolvedConfiguration().getFiles().containsAll([project(':subA').tasks["test"].binaryResultsDirectory.file("results.bin").get().getAsFile(),
```
Comment (big-guy) [inline subprojects/plugins/src/integTest/groovy/org/gradle/api/plugins/JvmTestSuitePluginIntegrationTest.groovy:258]: instead of reaching into other projects, it might be easier here to just manually specify the path.  e.g., just use `file("subA/build/test/results/results.bin")` (or whatever the right path is).  That'll let this test run with configuration caching and it's unlikely that we're going to be moving binaryResultsDirectory around in a way that makes this harder to maintain.

**Review by big-guy — CHANGES_REQUESTED** [review]

File: build-logic-commons/code-quality/src/main/kotlin/gradlebuild.code-quality.gradle.kts [hunk 23]
```diff
@@ -41,7 +41,7 @@ val rules by configurations.creating {
     isCanBeConsumed = false
 
     attributes {
-        attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, objects.named(LibraryElements.RESOURCES))
+        attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, objects.named(LibraryElements::class.java, LibraryElements.RESOURCES))
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply build-logic-commons/code-quality/src/main/kotlin/gradlebuild.code-quality.gradle.kts:44]: Sorry to pile-on but yes, I see lots of maintenance changes that appear orthogonal. Can we separate these into a different PR?

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java (hunk 5 above)
Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java]: I see why this would be confusing, I'll clarify the message.

File: build-logic-commons/code-quality/src/main/kotlin/gradlebuild.code-quality.gradle.kts (hunk 23 above)
[the remark this answers is not in the document]
Comment (tresat) [reply build-logic-commons/code-quality/src/main/kotlin/gradlebuild.code-quality.gradle.kts:44]: I think I agree with you at this point, Kyle.  I'll start a separate branch containing just the modification to `AttributeContainer` to just add the lazy overload and fix the fallout, and then we can hopefully merge that branch quickly and adjust this larger change to match.

File: subprojects/docs/src/snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/kotlin/producer/build.gradle.kts (hunk 16 above)
Comment (tresat) [reply subprojects/docs/src/snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/kotlin/producer/build.gradle.kts:18]: I'm just going to switch to using the 1-arg `objects.named` extension and remove this method completely.

File: build-logic-commons/code-quality/src/main/kotlin/gradlebuild.code-quality.gradle.kts (hunk 23 above)
[the remark this answers is not in the document]
Comment (tresat) [reply build-logic-commons/code-quality/src/main/kotlin/gradlebuild.code-quality.gradle.kts:44]: https://github.com/gradle/gradle/pull/18936

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java [hunk 24]
```diff
@@ -138,16 +197,31 @@ public boolean equals(Object o) {
...
     public int hashCode() {
         int result = parent != null ? parent.hashCode() : 0;
         result = 31 * result + state.hashCode();
+        result = 31 * result + lazyAttributes.hashCode();
         return result;
     }
+
+    private ImmutableAttributes evaluateLazyValues() {
+        final AttributeContainerInternal evaluatedAttributes = immutableAttributesFactory.mutable();
+        for (Map.Entry<Attribute<?>, Provider<?>> entry : lazyAttributes.entrySet()) {
```
[the remark this answers is not in the document]
Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java:220]: Still need a cast for key, but this simplifies it a little.

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java [hunk 25]
```diff
@@ -138,16 +197,31 @@ public boolean equals(Object o) {
...
+        }
+
         return state.equals(that.state);
     }
 
     @Override
     public int hashCode() {
         int result = parent != null ? parent.hashCode() : 0;
         result = 31 * result + state.hashCode();
+        result = 31 * result + lazyAttributes.hashCode();
```
[the remark this answers is not in the document]
Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java:214]: That definitely seems to be our hashCode pattern.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/LazyAttributesIntegrationTest.groovy [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+                    attributes {
+                        attribute(Usage.USAGE_ATTRIBUTE, testProvider.flatMap(value -> project.provider(() -> objects.named(Category, value))))
+                    }
+                }
+            }
+            """.stripIndent()
+
+        expect:
+        fails "outgoingVariants"
+        failure.assertHasCause("Unexpected type for attribute: 'org.gradle.usage'. Attribute type: org.gradle.api.attributes.Usage did not match actual type: org.gradle.api.attributes.Category\$Impl")
```
[the remark this answers is not in the document]
Comment (tresat) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/LazyAttributesIntegrationTest.groovy]: The message has expanded, so there's now text after that to verify.

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java [hunk 27]
```diff
@@ -138,16 +197,31 @@ public boolean equals(Object o) {
 
         DefaultMutableAttributeContainer that = (DefaultMutableAttributeContainer) o;
 
-        if (parent != null ? !parent.equals(that.parent) : that.parent != null) {
+        if (!Objects.equals(parent, that.parent)) {
             return false;
         }
+        if (!Objects.equals(lazyAttributes, that.lazyAttributes)) {
```
[the remark this answers is not in the document]
Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java:203]: I used evaluateLazyValues() instead and added a test.

**Comment by big-guy** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/core-api/src/main/java/org/gradle/api/attributes/AttributeContainer.java [hunk 28]
```diff
@@ -42,7 +42,6 @@
 @HasInternalProtocol
 @UsedByScanPlugin
 public interface AttributeContainer extends HasAttributes {
```
Comment (DPUkyle) [inline subprojects/core-api/src/main/java/org/gradle/api/attributes/AttributeContainer.java:44]: Only whitespace changes to this file; maybe revert it?

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultImmutableAttributes.java [hunk 29]
```diff
@@ -111,11 +111,6 @@ public int hashCode() {
         return hashCode;
     }
 
-    @Override
-    public <T> AttributeContainer attribute(Attribute<T> key, Provider<? extends T> provider) {
-        throw new UnsupportedOperationException("Mutation of attributes is not allowed");
-    }
-
```
Comment (DPUkyle) [inline subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultImmutableAttributes.java:118]: This is just moved to a different location in the file; maybe revert?

File: subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublicationsTest.groovy [hunk 30]
```diff
@@ -201,4 +207,36 @@ class DefaultConfigurationPublicationsTest extends Specification {
...
+        then:
+        prop.get().isEmpty()
+
+        when:
+        def file1 = new DefaultFileSystemLocation(new File("file1"))
+        prop.add(file1)
+
+        then:
+        prop.get().size() == 1
+        publications.getArtifacts().toSet()*.getName() == ["file1"]
```
Comment (DPUkyle) [inline subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublicationsTest.groovy:232]: `publications.artifacts*.name == ['file1']` can probably work here.

`DefaultConfigurationPublications#artifacts` returns `PublishArtifactSet`, which extends `Set`.

File: subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java [hunk 31]
```diff
@@ -868,6 +871,19 @@ public URI relativizeFrom(TestFile baseDir) {
...
+    /**
+     * Returns a human-readable relative path to this file from the base directory passed to create this TestFile.
+     *
+     * Fails if this TestFile was created in a way that did not provide a relative base.
+     *
+     * @see #relativizeFrom(TestFile)
+     * @see java.nio.file.Path#relativize(Path)
+     */
+    public String getRelativePathFromBase() {
+        Preconditions.checkArgument(!relativeBase.toPath().equals(this.toPath()), "relativeBase must have been set during construction");
```
Comment (DPUkyle) [inline subprojects/internal-testing/src/main/groovy/org/gradle/test/fixtures/file/TestFile.java:883]: I'm opposed to expanding the use of guava in our codebase. Since this is just a test, it's probably fine and this is not my hill to die on!

But also since this is a test, maybe a JUnit assertion is a better fit here?

File: subprojects/diagnostics/src/integTest/groovy/org/gradle/api/tasks/diagnostics/OutgoingVariantsReportTaskIntegrationTest.groovy [hunk 32]
```diff
@@ -39,6 +40,11 @@ class OutgoingVariantsReportTaskIntegrationTest extends AbstractIntegrationSpec
         run ':outgoingVariants'
 
         then:
+        def jarPath = new TestFile(getTestDirectory(), 'build/libs/myLib-1.0.jar').getRelativePathFromBase()
```
Comment (big-guy) [inline subprojects/diagnostics/src/integTest/groovy/org/gradle/api/tasks/diagnostics/OutgoingVariantsReportTaskIntegrationTest.groovy:43]: This can be simplified to just `file(...)`.  This will create the TestFile relative to the test directory for you

File: subprojects/docs/src/samples/build-organization/structuring-software-projects/groovy/build-logic/report-aggregation/src/main/groovy/com.example.report-aggregation.gradle [hunk 33]
```diff
@@ -44,17 +43,17 @@ def coverageDataPath = configurations.create('coverageDataPath') {
...
+        attribute(DocsType.DOCS_TYPE_ATTRIBUTE, objects.named(DocsType, DocsType.JACOCO_COVERAGE))
     }
 }
 
 // Register a code coverage report task to generate the aggregated report
 def codeCoverageReport = tasks.register('codeCoverageReport', JacocoReport) {
-    additionalClassDirs(classPath.filter { it.isDirectory() })
-    additionalSourceDirs(sourcesPath.incoming.artifactView { lenient(true) }.files)
-    executionData(coverageDataPath.incoming.artifactView { lenient(true) }.files.filter { it.exists() })
+    classDirectories.from(classesPath.getIncoming().getFiles())
```
Comment (big-guy) [inline subprojects/docs/src/samples/build-organization/structuring-software-projects/groovy/build-logic/report-aggregation/src/main/groovy/com.example.report-aggregation.gradle:54]: we need to keep the artifact view for now

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java [hunk 34]
```diff
@@ -100,6 +100,7 @@ private TestingFramework(Frameworks framework, String version) {
     private final JvmComponentDependencies dependencies;
     private boolean attachedDependencies;
     private final Action<Void> attachDependencyAction;
+    private final Property<String> testType = getObjectFactory().property(String.class);
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java:103]: You should be able to delete this

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java [hunk 35]
```diff
@@ -156,6 +157,11 @@ public DefaultJvmTestSuite(String name, ConfigurationContainer configurations, D
         });
     }
 
+    @Override
```
Comment (big-guy) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmTestSuite.java:160]: You should be able to delete this

**Review by big-guy — CHANGES_REQUESTED** [review]

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Review by big-guy — APPROVED** [review]

**Review by DPUkyle — APPROVED** [review]

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by big-guy** [comment]
@bot-gradle test and merge
